src/main/src/main/java/org/geoserver/security/decorators/DefaultSecureCatalogFactory.java

```diff
@@ -29,32 +38,125 @@ public boolean canSecure(Class clazz) {
+            return new SecuredCoverageInfo(unwrap((CoverageInfo) object), policy);
         else if (CoverageStoreInfo.class.isAssignableFrom(clazz))
-            return new SecuredCoverageStoreInfo((CoverageStoreInfo) object, policy);
+            return new SecuredCoverageStoreInfo(unwrap((CoverageStoreInfo) object), policy);
         else if (DataStoreInfo.class.isAssignableFrom(clazz))
-            return new SecuredDataStoreInfo((DataStoreInfo) object, policy);
+            return new SecuredDataStoreInfo(unwrap((DataStoreInfo) object), policy);
         else if (FeatureTypeInfo.class.isAssignableFrom(clazz))
-            return new SecuredFeatureTypeInfo((FeatureTypeInfo) object, policy);
+            return new SecuredFeatureTypeInfo(unwrap((FeatureTypeInfo) object), policy);
         else if (LayerInfo.class.isAssignableFrom(clazz))
-            return new SecuredLayerInfo((LayerInfo) object, policy);
+            return new SecuredLayerInfo(unwrap((LayerInfo) object), policy);
+        else if (WMSLayerInfo.class.isAssignableFrom(clazz))
+            return new SecuredWMSLayerInfo(unwrap((WMSLayerInfo) object), policy);
+        else if (WMTSLayerInfo.class.isAssignableFrom(clazz))
+            return new SecuredWMTSLayerInfo(unwrap((WMTSLayerInfo) object), policy);
         else
-            throw new IllegalArgumentException("Don't know how to wrap");
+            throw new IllegalArgumentException("Don't know how to wrap " + object);
     }
-
     /**
      * Returns {@link ExtensionPriority#LOWEST} since the wrappers generated by
      * this factory
+     * @return {@link ExtensionPriority#LOWEST}
      */
+    @Override
     public int getPriority() {
         return ExtensionPriority.LOWEST;
     }
 
+    private void logDoubleWrap(Object unwrapped, Object orig) {
+        String msg = String.format("Tried to double secure: %s already securing %s", orig, unwrapped);
+        LOGGER.warning(msg);
+    }
+
+    private void logDoubleWrap(final DataStoreInfo store) {
+        int proxies = 0;
+        int secured = 0;
+        DataStoreInfo object = store;
+        while (true) {
+            DataStoreInfo unwrapped;
+            if (object instanceof SecuredDataStoreInfo) {
+                secured++;
+                unwrapped = ((SecuredDataStoreInfo) object).unwrap(DataStoreInfo.class);
+            } else {
+                unwrapped = ModificationProxy.unwrap(object);
+                boolean isProxy = object != unwrapped;
+                if (isProxy) {
+                    proxies++;
+                } else {
+                    break;
+                }
+            }
+            object = unwrapped;
+        }
+
+        String msg = String.format("Double securing %s: proxies: %,d, SecuredDataStoreInfos: %,d",
+                store.getName(), proxies, secured);
+        LOGGER.warning(msg);
+    }
+
+    private WMTSLayerInfo unwrap(WMTSLayerInfo object) {
```

Comment:
`unwrap` methods that just return what they are given are a bit confusing.  If they are intended to always just be logging then maybe a different name would help.  Otherwise if they are more hooks to be used to deal with future bugs, maybe some documentation explaining your thinking.

Reply:
I've renamed the `unwrap` methods to `logIfSecured`, as that is what they do now.